File: ui/studio/src/models/studio.ts
export interface Connection<T> {
  pageInfo: PageInfo;
  edges: Edge<T>[];
  totalCount: number;
}

export interface Edge<T> {
  cursor: string;
  node: T;
}

export interface StudioGraph {
  mediaTypes: Connection<MediaType>;
  users: Connection<User>;
  stages: Connection<Stage>;
  tags: Connection<Tag>;
  media: Connection<Media>;
  whoami: User;
  notifications: Notification[];
}

export interface PageInfo {
  startCursor: string;
  endCursor: string;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}

export interface MediaType {
  id: string;
  name: string;
}

export interface User {
  id: string;
  dbId: number;
  displayName: string;
  username: string;
  roleName: string;
}

export interface Stage {
  id: string;
  name: string;
  dbId: number;
}

export interface Tag {
  id: string;
  name: string;
  dbId: number;
}

export interface AssignedStage {
  id: number
  name: string
  url: string
}

export type CopyrightLevel = 0 | 1 | 2 | 3;

export interface Media {
  id: string;
  name: string;
  src: string;
  sign: string;
  description: string;
  createdOn: string;
  size: number;
  copyrightLevel: CopyrightLevel;
  permissions: Permission[];
  assetType: MediaType;
  owner: User;
  stages: AssignedStage[];
  tags: string[];
  privilege: Privilege;
}

export interface MediaAttributes {
  multi: boolean;
  frames: string[];
  voice: Voice;
  link: Link;
  isRTMP: boolean;
  w: number;
  h: number;
}

export interface Link {
  blank: boolean;
  url: string;
}

export interface Voice {
}

export interface UploadFile {
  action?: string;
  filename?: string;
  data?: any;
  file: File;
  headers?: any;
  withCredentials?: boolean;
  method?: string;
  id: number;
  preview: string;
  status: 'local' | 'uploaded' | 'virtual';
  url?: string;
}

export interface Permission {
  id: string;
  approved: boolean;
  userId: number;
  assetId: number;
  seen: boolean;
  createdOn: string;
  note: null;
  user: User;
  asset: Media;
}

export interface Notification {
  id: string;
  type: "MEDIA_USAGE";
  mediaUsage: Permission;
}

export interface AvatarVoice {
  voice: string | null
  variant: string
  pitch: number
  speed: number
  amplitude: number
}

export type Privilege = 'NONE' | 'OWNER' | 'APPROVED' | 'PENDING_APPROVAL' | 'REQUIRE_APPROVAL'

File: core-v2/src/studios/graphql/studio.py
from ariadne import gql


type_defs = gql("""
    type User {
        id: ID!
        username: String!
        password: String!
        email: String!
        binName: String
        role: String!
        firstName: String!
        lastName: String!
        displayName: String
        active: Boolean!
        createdOn: String!
        firebasePushnotId: String
        deactivatedOn: String
        uploadLimit: Int
        intro: String
        canSendEmail: Boolean
        lastLogin: String
        roleName: String
    }
                
    enum AdminPlayerSortEnum {
        USERNAME_ASC
        USERNAME_DESC
        CREATED_ON_ASC
        CREATED_ON_DESC
        LAST_LOGIN_ASC
        LAST_LOGIN_DESC
        ROLE_ASC
        ROLE_DESC
        EMAIL_ASC
        EMAIL_DESC
    }

    type AdminPlayerConnection {
        totalCount: Int!
        edges: [User!]!
    }
                
    type Query {
        hello: String
        whoami: User
        adminPlayers(
            first: Int,
            page: Int,
            sort: [AdminPlayerSortEnum],
            usernameLike: String,
            createdBetween: [String]
        ): AdminPlayerConnection      
        media(input: MediaTableInput!): AssetConnection!  
        mediaList(mediaType: String, owner: String): [Asset!]!
    }      

    
    input MediaTableInput {
        page: Int
        limit: Int
        sort: [AssetSortEnum]
        name: String
        mediaTypes: [String]
        owners: [String]
        stages: [Int]
        tags: [String]
        createdBetween: [Date]
    }
                
    type Asset {
        id: ID!
        name: String
        src: String
        sign: String
        createdOn: Date
        size: Int!
        description: String
        assetType: AssetType
        owner: User
        stages: [Stage]
        tags: [String]
        copyrightLevel: Int
        permissions: [Permission]
        privilege: String
    }     

    type AssetConnection {
        totalCount: Int!
        edges: [Asset!]!
    }
                
    input BatchUserInput {
        username: String!
        password: String!
        email: String!            
    }
                
    type Permission {
        id: ID!
        userId: Int!
        assetId: Int!
        approved: Boolean!
        seen: Boolean!
        createdOn: Date!
        note: String
        user: User!
    }
                
    type AssetType {
        id: ID
        name: String!
    }

    type Stage {
        name: String!
        url: String!
        id: Int!
    }


    type Mutation { 
        batchUserCreation(users: [BatchUserInput]!, stageIds: [Int]): BatchUserCreationPayload
        uploadFile(base64: String!, filename: String!): File!
        saveMedia(input: SaveMediaInput!): SaveMediaPayload!
        deleteMedia(id: ID!): DeleteMediaPayload!
    }
                
    type BatchUserCreationPayload {
        users: [User!]!
    }
                
     fragment permissionFragment on Permission {
        id
        userId
        assetId
        approved
        seen
        createdOn
        note
        user {
            username
            displayName
        }
    }

    fragment f1 on User {
        id
        username
        password
        email
        binName
        role
        firstName
        lastName
        displayName
        active
        createdOn
        firebasePushnotId
        deactivatedOn
        uploadLimit
        intro
        canSendEmail
        lastLogin
        roleName
    }
                
    type File {
        url: String!
    }
                
    input SaveMediaInput {
        id: ID
        name: String!
        mediaType: String!
        copyrightLevel: Int!
        owner: String!
        stageIds: [Int!]!
        userIds: [Int!]
        tags: [String!]
        w: Float!
        h: Float!
        note: String
        urls: [String!]!
        voice: VoiceInput
        link: LinkInput
    }

    input VoiceInput {
        voice: String
        variant: String
        pitch: Int
        speed: Int
        amplitude: Int
    }

    input LinkInput {
        url: String
        blank: Boolean
        effect: Boolean
    }

    type SaveMediaPayload {
        asset: Asset!
    }
                
    type DeleteMediaPayload {
        success: Boolean!
        message: String!
    }

    scalar Date

    enum AssetSortEnum {
        ASSET_TYPE_ID_ASC
        ASSET_TYPE_ID_DESC
        OWNER_ID_ASC
        OWNER_ID_DESC
        NAME_ASC
        NAME_DESC
        CREATED_ON_ASC
        CREATED_ON_DESC
    }
""")
